From d2efa611fcc35baf25329a32ff84fad26f078a83 Mon Sep 17 00:00:00 2001
From: Alexey Kodanev <aleksei.kodanev@bell-sw.com>
Date: Wed, 27 Nov 2024 15:22:32 +0000
Subject: [PATCH] ldd: detect static-pie binaries

Currently, ldd prints confusing information about static pie
binaries. For example, when using LD_PRELOAD:

	apk add mimalloc-global
	source /etc/profile.d/mimalloc.sh
	echo "int main(void) {return 0;}" | gcc -static-pie -o test -x c -

	ldd test
	        /usr/lib/ld-musl-aarch64.so.1 (0xffff9f393000)
	        /lib/libmimalloc.so.2 => /lib/libmimalloc.so.2 (0xffff9f30c000)
	        libc.musl-aarch64.so.1 => /usr/lib/ld-musl-aarch64.so.1 (0xffff9f393000)

It is not at all obvious that this is a statically linked
binary that does not use or depend on the libs listed above.

With the patch, the output will be as follows:

	ldd test
		statically linked (static-pie)
---
 ldso/dynlink.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/ldso/dynlink.c b/ldso/dynlink.c
index 90fca727..95426fce 100644
--- a/ldso/dynlink.c
+++ b/ldso/dynlink.c
@@ -1960,6 +1960,17 @@ void __dls3(size_t *sp, size_t *auxv)
 				if (app.phdr[i].p_type == PT_INTERP)
 					ldso.name = laddr(&app, app.phdr[i].p_vaddr);
 			}
+			int needed_so = 0;
+			for (i=0; app.dynv[i]; i+=2) {
+				if (app.dynv[i]==DT_NEEDED) {
+					needed_so = 1;
+					break;
+				}
+			}
+			if (!needed_so) {
+				dprintf(1, "\tstatically linked (static-pie)\n");
+				_exit(0);
+			}
 			dprintf(1, "\t%s (%p)\n", ldso.name, ldso.base);
 		}
 	}
-- 
2.25.1

